controllers/namespace/namespace_controller.go

```diff
@@ -0,0 +1,197 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// once is a pointer to the sync.Once type
+var once sync.Once
+
+// NamespaceReconciler reconciles the namespaces starting with openshift-*
+// nolint:revive
+type NamespaceReconciler struct {
+	client.Client
+	ctx    context.Context
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=csiaddons.openshift.io,resources=pods,verbs=list;delete
+// +kubebuilder:rbac:groups=core,resources=namespaces,verbs=get;list;watch;update;patch
+
+// Reconcile reads that state of the cluster for a Namespace object and makes changes based on the state read
+// i.e,the annotations of the namespace
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *NamespaceReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	prevLogger := r.Log
+	defer func() { r.Log = prevLogger }()
+	r.Log = r.Log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	r.ctx = ctx
+
+	r.Log.Info("Reconciling Namespace.", "Namespace", klog.KRef(request.Namespace, request.Name))
+
+	var err error
+	updated := false
+
+	once.Do(func() {
+		updated, err = r.annotateAllOpenshiftNamespaces()
+	})
+
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	ns := &corev1.Namespace{}
+	err = r.Client.Get(ctx, request.NamespacedName, ns)
+	if err != nil {
+		r.Log.Error(err, fmt.Sprintf("Namespace %s does not exist", request.Name))
+		return reconcile.Result{}, err
+	}
+
+	// Annotate namespaces starting with openshift-*, but ignore namespace if
+	// it has annotation "'skipReclaimspaceSchedule':'true'" set
+	updated, err = r.annotateOpenshiftNamespaces([]corev1.Namespace{*ns})
+	if err != nil {
+		r.Log.Error(err, "Failed annotating openshift-* namespaces")
+		return reconcile.Result{}, err
+	}
+
+	// Restart the csi-addons controller pod if any namespace was annotated
+	if updated {
+		podList := &corev1.PodList{}
+		listOpts := []client.ListOption{
+			client.InNamespace(os.Getenv(util.OperatorNamespaceEnvVar)),
+			client.MatchingLabels{"app.kubernetes.io/name": "csi-addons"},
+		}
+		err := r.Client.List(ctx, podList, listOpts...)
+		if err != nil {
+			r.Log.Error(err, "Failed to get list of pods with label 'csi-addons'")
+			return reconcile.Result{}, err
+		}
+
+		// If no csi-addons pod is present
+		if len(podList.Items) == 0 {
+			return reconcile.Result{}, nil
+		}
+
+		for _, csiAddonsPod := range podList.Items {
+			err = r.Client.Delete(ctx, &csiAddonsPod, client.GracePeriodSeconds(0))
+			if err != nil {
+				r.Log.Error(err, "Failed to delete csi-addons pod")
+				return reconcile.Result{}, err
+			}
+		}
```

Comment:
Let's make this as a separate func, Which can be called twice here and in once.DO, Because your namespace will be already updated here and it won't go into this condition at all.

Reply:
Yeah, right.
Updated the code to encapsulate this code within a separate function.